### DISCUSSION TITLE
Help Needed to Interpret Results

### DISCUSSION THREAD

**Discussion by sqrk** [NVIDIA-NeMo/NeMo#5287] [Q&A] [answered]
Hello!
Thank you once again for such a great toolkit.
I would really appreciate it if @titu1994 or someone else could help me interpreting these results.

I trained a couple of architectures on an impaired speech dataset (relatively small, around 60h total). It has very high speaker variability due to the speech disabilities and has different severity levels (the dataset is not perfectly balanced across the different levels). The speech is very hard to understand to humans too so high error rates are expected. The current SOTA for this task is around 25% WER. The dataset only contains single words. All of the models were fine-tuned using the available pre-trained checkpoints. For the sub-word models, I used a vocabulary size of 1024.

The results are as follows:

| Model | WER (%) | CER (%) |
| -------------| ---------| ---------|
| Conformer CTC (sm) | **47.79** | 48.68 |
| Jasper | 50.93 | **36.77** |
| Conformer Transducer (sm) | 51.29 | 52.62 |
| QuartzNet | 51.86 | 54.03 |
| CitriNet (1024) | 58.54 | 54.40 |

1) This was surprising to me because I was expecting auto-regressive models (Conformer Transducer and CitriNet) to perform better that CTC ones. All the information that I found about Transducer loss vs CTC loss stated that Transducer loss gives better accuracy but has slower convergence and a more complex training. Is there anything else that can make CTC models perform better? Could it be that since I'm using a single-word dataset, there isn't much to gain from context (because the sequences are relatively short)?

2) The second surprising thing was that Jasper was better than QuartzNet, which itself was better than CitriNet (even though they're all improvements of the previous one). I thought that it could be due to Jasper having much more parameters but for the Conformers, the smaller version performed better than the larger ones (albeit by less than 2% absolute difference).

3) The Conformer CTC performs better than Jasper for speakers with very low intelligibility but much worse for speakers with high intelligibility. For example, here are the results fine-tuned on the data of 1) the "best" speaker in the high intelligibility category and 2) the "worst" speaker in the low one:

| Model | Speaker 1 WER | Speaker 2 WER | 
|-----|----|----|
| Conformer CTC | 40.35 | 43.91 |
| Jasper | 15.11 | 62.03 |

One theory that I thought of is the fact that Jasper is character-based and the Conformer uses a sub-word tokenizer. Using sub-words might be introducing a lot of mistakes in the higher intelligibility levels but "saving the day" when the speech is very troubled (so although part of the subword is mis-pronounced, the larger token helps forgiving some of the mispronunciation).

4) When testing on another dataset (to assess the robustness to unseen speakers. This dataset hasn't been used during training), Jasper performed much better than the Conformer:

| Model | WER | CER |
|-|-|-|
| Conformer CTC | 95.82 | 81.10 |
| Jasper | 57.27 | 32.62 |

There is such a huge disparity between the two. This second dataset includes some sentences. After removing them the Conformer improved slightly (86% WER) and when I looked at the predictions I found that it was only predicting single words even when met with a sentence.  What could this be due to? It was pre-trained on sentences after all. Jasper was predicting the sentences correctly, and even has the CER increase when the sentences are removed. Does the context help Jasper's predictions in any way? As far as I understood, CTC models only base themselves off of the acoustics (unless you add a language model).

5) Could the CNN-based architecture of Jasper be helping the accuracy in any way?

Thank you so much, and sorry for the long post!

**Comment by titu1994** [comment, marked as answer]
Thanks for the post ! Let's break it down step by step since there's many parts 

> trained a couple of architectures on an impaired speech dataset (relatively small, around 60h total)

Can you give training details - did you change Tokenizer, what number of epochs, LR and other details ? CTC models are *slower* to converge than RNNT models. 

> The dataset only contains single words.

This kind of dataset is very challenging for ASR models that were pretrained on entire sentences. They have been tuned to work with long utterances and therefore need lots of context for correct predictions. This occurs more so for RNNT even because both acoustic and LM model want more context. 

 > 1. This was surprising to me because I was expecting auto-regressive models (Conformer Transducer and CitriNet) to perform better that CTC ones. All the information that I found about Transducer loss vs CTC loss stated that Transducer loss gives better accuracy but has slower convergence and a more complex training. Is there anything else that can make CTC models perform better? Could it be that since I'm using a single-word dataset, there isn't much to gain from context (because the sequences are relatively short)?

Citrinet is non autoregressive, similar to conformer CTC. RNNT training might deteriorate due to short context. Have to tried enabling fastemit loss ? You can do model.loss._loss.fastemit_lambda = 0.001 as a temporary measure. Or simply create a new model with correct config and use new_model.load_state_dict(old_model.state_dict())

RNNT always has faster convergence speed than CTC and can therefore overfit faster as well. 

> 2. The second surprising thing was that Jasper was better than QuartzNet, which itself was better than CitriNet (even though they're all improvements of the previous one). I thought that it could be due to Jasper having much more parameters but for the Conformers, the smaller version performed better than the larger ones (albeit by less than 2% absolute difference)

This indicates you're lacking the parameter count for sufficient training. Jasper is some 150+ M params or more, QN is 19. Yes, QN can get better scores but it requires careful tuning. You might want to use Conformer CTC Large (120 M) to have reasonable comparison with Jasper. 

I am also confused as to how CER can be lower for Jasper but WER is higher. That points to very noisy ground truth that doesn't correspond to what the model is predicting. Think of cases where the user says hmm or umm, model predicts hmm or umm correctly but ground truth does not have those words. 

> 3. The Conformer CTC performs better than Jasper for speakers with very low intelligibility but much worse for speakers with high intelligibility. For example, here are the results fine-tuned on the data of 1) the "best" speaker in the high intelligibility category and 2) the "worst" speaker in the low one:

This can be either due to model capacity - Conformer small is just 10 M params, it is not augmented with noise augmentation either. Or it can point to the 4x and 8x models doing worse for single word dataset. This corresponds with the fact that Citrinet with 8x has the worst wer and cer despite similar param count to Jasper, which has just 2x stride. 

> 4. There is such a huge disparity between the two. This second dataset includes some sentences. After removing them the Conformer improved slightly (86% WER) and when I looked at the predictions I found that it was only predicting single words even when met with a sentence. What could this be due to? It was pre-trained on sentences after all. Jasper was predicting the sentences correctly, and even has the CER increase when the sentences are removed. Does the context help Jasper's predictions in any way? As far as I understood, CTC models only base themselves off of the acoustics (unless you add a language model).

Pretrained info was entirely lost the moment you change Tokenizer + fine tune on your dataset. The model has entirely forgotten about it's entire original dataset due to one word training - it is only capable of that now. 

Look at the following paper that discusses finetuning on Google Speech Commands - similar one word Dataset. The model immediately starts to forget the original training results and gets extremely high wer on original dataset after just 1 epoch of training. 

https://arxiv.org/abs/2210.03255

That Jasper continues to work is that it is character based model - it did not need to change vocabulary and so did not forget original training. Or that 4x stride very much hurts this kind of 1 word training. Can't be sure what's the cause. 

Yes, CTC models only base predictions on the acoustic model. But due to their size and global context they do model text to some degree. 

> 5. Could the CNN-based architecture of Jasper be helping the accuracy in any way?

If cnn alone would help, Citrinet would do very well too, along with QuartzNet. All of them are purely convolutional models. But apparently they don't.

Finally, is it really necessary to change the Tokenizer ? Does the text in the new dataset have tokens not in the original Tokenizer ? You might be able to get much better results by normalizing the ground truth of your dataset to match the Tokenizer and then training a model with full weights without changing tokenizer.  

Plus, even if you must change Tokenizer, make sure to load the weights of the older model back into new model - all 3 components for RNNT (encdecjoint) and 2 components for CTC. (EncDec)

**Reply by sqrk** [reply]
Thank you so much for your response!

> Can you give training details - did you change Tokenizer, what number of epochs, LR and other details ? CTC models are slower to converge than RNNT models.

Tokenizer: I changed the vocabulary for the sub-word models to use my dataset's vocabulary but kept the size at 1024. 
Epochs: Around 20. They converge around the 9th though (and around 5th for Conf-T).
LR:
| Model | Best LR Rec. |
|-|-|
| Conf-T | 5 |
| Conf-CTC | 1 | 
| QuartzNet | 1e-4 |
| Jasper | 1e-5 |
| Citrinet | 1e-5 |

I kept the same optimizer, scheduler, SpecAug...

> Have to tried enabling fastemit loss ?

No, but I'll give it a try. Thanks!

> You might want to use Conformer CTC Large (120 M) to have reasonable comparison with Jasper.

The large Conformers actually performed worse than the small versions although with less than 1% absolute difference.

> Think of cases where the user says hmm or umm, model predicts hmm or umm correctly but ground truth does not have those words.

I can imagine this happening because sometimes there's stuttering in the speech and false starts. But this would show in the predictions right? From what I've seen all the predictions are single words but I've noticed mistakes like:

```
reference:anybody
predicted:anyway

reference:will
predicted:wil

reference:plane
predicted:plain

reference:allure
predicted:alure
```
I'm assuming that adding an LM would fix this.

> This corresponds with the fact that Citrinet with 8x has the worst wer and cer

Is this stride the same as the Conv layer's stride (which is of 2 according to the paper)? I found the paper mentioning a "8x compression in the time domain", but I'm not sure if it's the same concept (sorry I'm very new to the field). I also can't find where it is achieved in the config file (I was expecting some parameter stride: 8, but I guess it's more complex than that).
So if we apply a high stride on a short sequence (as is the case with single word utterances), it ends up missing information? Or is there another relationship?

> https://arxiv.org/abs/2210.03255

Thank you so much, this is very useful.

> Yes, CTC models only base predictions on the acoustic model. But due to their size and global context they do model text to some degree.

Do you have any resource in mind where I can learn about how this is achieved?

Thank you so much!

**Reply by titu1994** [reply]
Check your ground truth text to see if you need to change tokenizer, that is severely degrading the wer. CTC models make spelling mistakes, yes LM helps. 

Rnnt can do it without LM but apparently not for very difficult speech. 

There are 3 stride layers in Citrinet, each with 2x stride. So overall 8x reduction in stride. Jasper has just 2x stride at the beginning. 

For most acoustic models, high stride is necessary to be efficient on sentence level speech. For single word, stride hurts cause the model does not have sufficient context to correctly predict tokens.

There's no resource as such, it's more of an empirical finding from the results of Jasper and Citrinet.
